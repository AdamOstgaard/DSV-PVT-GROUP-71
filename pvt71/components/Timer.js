import React from "react";
import { AppRegistry, Text, View, StyleSheet } from "react-native";
import moment from "moment";
import "moment-duration-format";
import { AsyncStorage } from "react-native";

export class Timer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      timerRunning: false,
      time: this.props.startTime,
      counter: 0,
      sleepTime: {
        start: null,
        end: null
      },
      sleeping: false,
      stopsleep: false
    };
  }

  componentDidMount() {
    this.getSettings();
    this.startTimer(this.state.time);
    //console.log("sleeping: " + this.state.sleeping)
  }

  async componentDidUpdate(prevProps, prevState) {
    if (
      prevState.sleepTime.start !== this.state.sleepTime.start ||
      prevState.sleepTime.end !== this.state.sleepTime.end
    ) {
      this.setState({ stopsleep: false });
    }

    //console.log("sleepin is " + this.state.sleeping);
    var hours = new Date().getHours(); //Current Hours
    var min = new Date().getMinutes(); //Current Minutes
    var currentTimeInMilliseconds = this.toMilliseconds(hours, min);
    if (prevProps.onFocus !== this.props.onFocus) {
      await this.getSettings();
      //console.log("sleeping1: " + this.state.sleeping);
    }
    if (
      currentTimeInMilliseconds >= this.state.sleepTime.start &&
      currentTimeInMilliseconds <= this.state.sleepTime.end
    ) {
      if (
        !this.state.sleeping &&
        this.state.sleepTime.start !== null &&
        !this.state.stopsleep
      ) {
        //console.log("sleeping2: " + this.state.sleepTime.end);
        this.setState({ sleeping: true }, () => {
          this.pauseTimer();
          this.startTimer(
            this.state.sleepTime.end - this.state.sleepTime.start
          );
          this.props.onTimerSleep(true);
        });
      }
    } else if (
      currentTimeInMilliseconds >= this.state.sleepTime.start &&
      currentTimeInMilliseconds >= this.state.sleepTime.end
    ) {
      if (
        !this.state.sleeping &&
        this.state.sleepTime.start !== null &&
        !this.state.stopsleep
      ) {
        //console.log("sleeping2: " + this.state.sleepTime.end);
        this.setState({ sleeping: true }, () => {
          this.pauseTimer();
          this.startTimer(
            this.toMilliseconds(24,0) - this.state.sleepTime.start + this.state.sleepTime.end
          );
          this.props.onTimerSleep(true);
        });
      }
    }
  }

  toMilliseconds = (h, m) =>
    moment.duration(h, "h").asMilliseconds() +
    moment.duration(m, "m").asMilliseconds();

  async getSettings() {
    try {
      const newSleepTime = await AsyncStorage.getItem("sleeptime");
      const s = JSON.parse(newSleepTime);
      const start = s.startTime;
      const end = s.endTime;
      //console.log(end);
      await this.setState({ sleepTime: { start, end } });
    } catch (error) {
      console.log("ERROR");
    }
  }

  render() {
    return (
      <View style={styles.timerContainer}>
        <Text
          {...this.props}
          onPress={() => {
            if (this.props.onPress) this.props.onPress();
            this.props.onReset(false);
            this.resetTimer();
          }}
          style={styles.timerText}
        >
          {this.formatTime(this.state.time)}
        </Text>
        <Text style={{ fontSize: 12, borderWidth: 0, display: "flex" }}>
          Press to reset
        </Text>
      </View>
    );
  }

  resetTimer() {
    if (this.state.sleeping) {
      return;
    }
    this.props.onReset(false);
    this.stopTimer();
    this.startTimer(this.props.startTime);
  }

  startTimer(time) {
    this.setState({ time, timerRunning: true });
    const timerInterval = 1000;

    this.timer = setInterval(
      () => this.handleTick(timerInterval),
      timerInterval
    );
  }

  handleTick(interval) {
    this.setState({
      time: this.state.time - interval,
      counter: this.state.counter + 1
    });

    if (this.state.counter === 60) {
      this.setState({ counter: 0 });
    }

    if (this.state.time > 0) {
      return;
    }
    if (this.state.sleeping) {
      this.setState({ sleeping: false, stopsleep: true });
      this.stopTimer();
      this.startTimer(this.props.startTime);
      this.props.onTimerSleep(false);
      //console.log("vilan är över");
      return;
    }

    if (this.props.callback && !this.state.sleeping) {
      this.props.callback(() => this.resetTimer());
    }

    this.pauseTimer();
  }

  pauseTimer() {
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.setState({ timerRunning: false });
  }

  stopTimer() {
    this.pauseTimer();
    this.setState({ timerRunning: false, time: this.props.startTime });
  }

  resumeTimer() {
    startTimer(this.state.time);
  }

  formatTime = time => moment.duration(time).format("h:mm:ss");
}
const styles = StyleSheet.create({
  timerContainer: {
    justifyContent: "center",
    alignItems: "center",
    textAlign: "center",
    textAlignVertical: "center",
    width: 280,
    height: 280,
    borderStyle: "solid",
    borderWidth: 2,
    borderRadius: 1000,
    flexDirection: "column"
  },
  timerText: {
    fontSize: 64,
    display: "flex"
  }
});

AppRegistry.registerComponent("Timer", () => Timer);
