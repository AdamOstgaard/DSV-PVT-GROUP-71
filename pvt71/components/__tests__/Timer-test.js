import 'react-native';
import React from 'react';
import { Timer } from '../Timer';
import renderer from 'react-test-renderer'

describe('Timer Test', () => {
    describe('Timer Snapshot', () =>{
        it('Renders correctly', () => {
            const tree = renderer.create(<Timer />).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
    
    describe('Timer start', () => {
        it('timer start successful', () => {
            let timerData = renderer.create(<Timer/>).getInstance();
            timerData.startTimer(10);
            expect(timerData.state.timerRunning).toEqual(true);
            expect(timerData.state.time).toEqual(10);
            timerData.stopTimer();
        });
    });

    describe('Timer pause', () => {
        it('Timer stop successful', () => {
            let timerData = renderer.create(<Timer/>).getInstance();
            timerData.startTimer(10);
            timerData.pauseTimer();
            expect(timerData.state.timerRunning).toEqual(false);
            expect(timerData.state.time).toEqual(10);
        });
    });

    describe('Callback works correctly', () => {
        it('Calls callback on timer end', () => {
            let success = false;
            let timerData = renderer.create(<Timer callback={() => {
                console.log('Test');
                success = true;}}/>).getInstance();
            timerData.startTimer(10);
            timerData.handleTick(99999);
            expect(success).toEqual(true);
        });
    });


    describe('Timer state', () => {
        it('state correct', () => {
            let timerData = renderer.create(<Timer/>).getInstance();
            timerData.startTimer(10);
            expect(timerData.state.timerRunning).toEqual(true);
            timerData.pauseTimer();
            expect(timerData.state.timerRunning).toEqual(false);
            timerData.startTimer(10);
            expect(timerData.state.timerRunning).toEqual(true);
            timerData.stopTimer();
            expect(timerData.state.timerRunning).toEqual(false);
        });
    });

    describe('resetTimer', () => {
        it('resetTimer works correctly', () => {
            let timerData = renderer.create(<Timer/>).getInstance();
            timerData.startTimer();
            timerData.stopTimer();
            expect(timerData.state.timerRunning).toEqual(false);
            timerData.resetTimer();
            expect(timerData.state.timerRunning).toEqual(true);
            timerData.stopTimer();
        });
    });

    describe('Handle tick', () => {
        it('Handles tick correctly', () => {
            let timerData = renderer.create(<Timer/>).getInstance();
            timerData.startTimer(1100);
            timerData.handleTick(1000);
            expect(timerData.state.time).toEqual(100);
            timerData.handleTick(100);
            timerData.handleTick(1000);
            expect(timerData.state.timerRunning).toEqual(false);
        });
    });

});