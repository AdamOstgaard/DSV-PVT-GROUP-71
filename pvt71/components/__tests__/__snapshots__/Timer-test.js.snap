// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Test Timer Snapshot Renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 1000,
      "borderStyle": "solid",
      "borderWidth": 2,
      "flexDirection": "column",
      "height": 280,
      "justifyContent": "center",
      "textAlign": "center",
      "textAlignVertical": "center",
      "width": 280,
    }
  }
>
  <Text
    onPress={[Function]}
    style={
      Object {
        "display": "flex",
        "fontSize": 64,
      }
    }
  >
    0
  </Text>
  <Text
    style={
      Object {
        "borderWidth": 0,
        "display": "flex",
        "fontSize": 12,
      }
    }
  >
    Press to reset
  </Text>
</View>
`;
